### PR TITLE
refactor: break out service client runtime into own module

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

N/A
## Description of changes

Split off client only runtime components into separate module from `runtime-core`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by lauzadis — APPROVED** [review]

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt [hunk 1]
```diff
@@ -186,6 +173,19 @@ object RuntimeTypes {
         }
     }
 
+    object SmithyClient : RuntimeTypePackage(KotlinDependency.SMITHY_CLIENT) {
```
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/RuntimeTypes.kt:176]: **Question**: Same question about naming as with module name.

File: tests/codegen/paginator-tests/build.gradle.kts [hunk 2]
```diff
@@ -65,6 +65,7 @@ dependencies {
 
     compileOnly(project(":smithy-kotlin-codegen"))
     implementation(project(":runtime:runtime-core"))
+    implementation(project(":runtime:smithy-client"))
```
Comment (ianbotsf) [inline tests/codegen/paginator-tests/build.gradle.kts:68]: **Question**: Do we need to explicitly depend on **runtime-core** if **smithy-client** pulls it in as an API dependency?

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/Exceptions.kt [hunk 3]
```diff
@@ -82,12 +81,6 @@ public open class ServiceErrorMetadata : ErrorMetadata() {
         public val RequestId: AttributeKey<String> = AttributeKey("RequestId")
     }
 
-    /**
-     * The name of the service that sent this error response
-     */
-    public val serviceName: String
-        get() = attributes.getOrNull(SdkClientOption.ServiceName) ?: ""
-
```
Comment (ianbotsf) [inline runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/Exceptions.kt:90]: **Question**: Why was this removed?

File: runtime/smithy-client/common/src/aws/smithy/kotlin/runtime/client/SdkClientOption.kt [hunk 4]
```diff
@@ -31,6 +31,11 @@ public object SdkClientOption {
      * The client logging mode (see [SdkLogMode]
      */
     public val LogMode: ClientOption<SdkLogMode> = ClientOption("LogMode")
+
+    /**
+     * Whether endpoint discovery is enabled or not. Default is true
+     */
+    public val EndpointDiscoveryEnabled: ClientOption<Boolean> = ClientOption("EndpointDiscoveryEnabled")
```
Comment (ianbotsf) [inline runtime/smithy-client/common/src/aws/smithy/kotlin/runtime/client/SdkClientOption.kt:38]: **Question**: Why was this added?

File: settings.gradle.kts [hunk 5]
```diff
@@ -51,6 +51,7 @@ include(":runtime:serde")
 include(":runtime:serde:serde-form-url")
 include(":runtime:serde:serde-json")
 include(":runtime:serde:serde-xml")
+include(":runtime:smithy-client")
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply settings.gradle.kts:54]: client isn't very specific though either we have multiple types of clients (e.g. http client). Smithy client was to signify it's applicable to generated smithy clients (just like a hypothetical `smithy-server` would be applicable to generated servers)

File: tests/codegen/paginator-tests/build.gradle.kts (hunk 2 above)
Comment (aajtodd) [reply tests/codegen/paginator-tests/build.gradle.kts:68]: Technically no but it harms nothing to have it this way.

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/Exceptions.kt (hunk 3 above)
Comment (aajtodd) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/Exceptions.kt:90]: Because I moved `SdkClientOption` with the rest of the client stuff out of core. I looked through codegen, runtime, and generated clients. We don't ever populate this. I think we ought to remove it given that all generated exceptions inherit from a service specific exception already. I can be talked into a different direction though. I have questions/feels in general about `SdkClientOption` right now.

File: runtime/smithy-client/common/src/aws/smithy/kotlin/runtime/client/SdkClientOption.kt (hunk 4 above)
Comment (aajtodd) [reply runtime/smithy-client/common/src/aws/smithy/kotlin/runtime/client/SdkClientOption.kt:38]: Moved from `aws-sdk-kotlin`. We can remove it right now completely since it's unused but it was solving a FIXME from `aws-sdk-kotlin`.
